apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs

```diff
@@ -0,0 +1,311 @@
+                            s_sessionsCreating++;
+                            return;
+                        }
+                    }
+                }
+                else
+                {
+                    throw new RpcException(new Status(StatusCode.ResourceExhausted, "Number of available Sessions exhausted."));
+                }
+            }
+        }
+
+        private static void EndSessionCreating()
+        {
+            lock (s_waitSync)
+            {
+                s_sessionsCreating--;
+            }
+            SignalAnyWaitingRequests();
+        }
+
+        /// <summary>
+        /// Allocates a session from the pool if possible, or creates a new Spanner Session.
+        /// </summary>
+        /// <param name="spannerClient"></param>
+        /// <param name="project"></param>
+        /// <param name="spannerInstance"></param>
+        /// <param name="spannerDatabase"></param>
+        /// <param name="options"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public static async Task<Session> CreateSessionFromPoolAsync(this SpannerClient spannerClient,
+            string project, string spannerInstance, string spannerDatabase, TransactionOptions options, CancellationToken cancellationToken)
+        {
+            project.ThrowIfNullOrEmpty(nameof(project));
+            spannerInstance.ThrowIfNullOrEmpty(nameof(spannerInstance));
+            spannerDatabase.ThrowIfNullOrEmpty(nameof(spannerDatabase));
+            
+            await StartSessionCreating(cancellationToken).ConfigureAwait(false);
+            try
+            {
+                var sessionPoolKey = new SessionPoolKey {
+                    Client = spannerClient,
+                    Project = project,
+                    Instance = spannerInstance,
+                    Database = spannerDatabase
+                };
+                SessionPoolImpl targetPool = GlobalPool.GetOrAdd(sessionPoolKey, key => new SessionPoolImpl(key));
+
+                var session = await targetPool.AcquireSessionAsync(options, cancellationToken);
+                if (session != null)
+                {
+                    //refresh the mru list which tells us what sessions need to be trimmed from the pool when we want
+                    // to add another poolEntry.
+                    ReSortMru(targetPool);
+                    s_sessionsInUse.AddOrUpdate(session, x => sessionPoolKey, (x, y) => sessionPoolKey);
+                    LogSessionsInUse();
+                }
+                return session;
+            }
+            finally
+            {
+                EndSessionCreating();
+            }
+        }
+
+        private static void SignalAnyWaitingRequests()
+        {
+            TaskCompletionSource<int> waitingSessionRequest;
+            //if anyone is waiting, let them query for their session.
+            if (CurrentActiveSessions < MaximumActiveSessions
```

Comment:
Any reason to make this conditional rather than just relying on the queue? I suspect it'll be easier to prove to ourselves that we don't have any race conditions, the fewer things like this we have.